openedx/core/lib/block_cache/block_structure.py

```diff
@@ -0,0 +1,588 @@
+        # that are still linked
+        pruned_block_relations = defaultdict(_BlockRelations)
+        old_block_relations = self._block_relations
+
+        # Build the structure from the leaves up by doing a post-order
+        # traversal of the old structure, thereby encountering only
+        # reachable blocks.
+        for block_key in self.post_order_traversal():
+            # If the block is in the old structure,
+            if block_key in old_block_relations:
+                # Add it to the new pruned structure
+                self._add_block(pruned_block_relations, block_key)
+
+                # Add a relationship to only those old children that
+                # were also added to the new pruned structure.
+                for child in old_block_relations[block_key].children:
+                    if child in pruned_block_relations:
+                        self._add_to_relations(pruned_block_relations, block_key, child)
+
+        # Replace this structure's relations with the newly pruned one.
+        self._block_relations = pruned_block_relations
+
+    def _get_block_keys(self):
+        """
+        Returns an iterator of all the block keys in the block
+        structure.
+        """
+        return self._block_relations.iterkeys()
+
+    def _add_relation(self, parent_key, child_key):
+        """
+        Adds a parent to child relationship in this block structure.
+
+        Arguments:
+            parent_key (UsageKey) - Usage key of the parent block.
+            child_key (UsageKey) - Usage key of the child block.
+        """
+        self._add_to_relations(self._block_relations, parent_key, child_key)
+
+    @staticmethod
+    def _add_to_relations(block_relations, parent_key, child_key):
+        """
+        Adds a parent to child relationship in the given block
+        relations map.
+
+        Arguments:
+            block_relations (defaultdict({UsageKey: _BlockRelations})) -
+                Internal map of a block's usage key to its
+                parents/children relations.
+
+            parent_key (UsageKey) - Usage key of the parent block.
+
+            child_key (UsageKey) - Usage key of the child block.
+        """
+        block_relations[child_key].parents.append(parent_key)
+        block_relations[parent_key].children.append(child_key)
+
+    @staticmethod
+    def _add_block(block_relations, usage_key):
+        """
+        Adds the given usage_key to the given block_relations map.
+
+        Arguments:
+            block_relations (defaultdict({UsageKey: _BlockRelations})) -
+                Internal map of a block's usage key to its
+                parents/children relations.
+
+            usage_key (UsageKey) - Usage key of the block that is to
+                be added to the given block_relations.
+        """
+
+        # Since block_relations is a DefaultDict, the usage_key will
+        # automatically be added by virtue of accessing it.
+        _ = block_relations[usage_key]
```

Comment:
Do you even need _add_block if block_relations is a defaultdict?  Seems like _add_to_relations will add them anyway.  Also, this is a weirdly unintuitive use of the defaulting behavior.  Might be clearer to either assign `block_relations[usage_key] = _BlockRelations()` explicitly, or for better encapsulation make `block_relations` a dict/defaultdict subclass with an `add_block(usage_key)` method on it.  

Also, it's a little confusing that we have a class called `_BlockRelations`, and a collection of those relations that is also called `block_relations`.  Maybe the collection could be something like `self._block_relations_mapping`.